### PR TITLE
TIP-1073: Duplicate ReadValueFactoryCollection for Write purpose

### PR DESCRIPTION
**Pull request by anaelChardan**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | -
| Added legacy Behats               | Todo
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Component/Product/Factory/Read/Value/OptionValueFactory.php [hunk 1]
```diff
@@ -34,6 +35,24 @@ public function create(Attribute $attribute, ?string $channelCode, ?string $loca
         return OptionValue::value($attributeCode, $data);
     }
 
+    public function createByCheckingData(Attribute $attribute, ?string $channelCode, ?string $localeCode, $data): ValueInterface
+    {
+        if (null === $data) {
+            return $this->createWithoutCheckingData($attribute, $channelCode, $localeCode, $data);
+        }
+
+        //QUESTION: SHALL WE TEST IF IT EXISTS ?
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Factory/Read/Value/OptionValueFactory.php]: to remove

File: src/Akeneo/Pim/Enrichment/Component/Product/Factory/Read/Value/OptionsValueFactory.php [hunk 2]
```diff
@@ -35,6 +37,27 @@ public function create(Attribute $attribute, ?string $channelCode, ?string $loca
         return OptionsValue::value($attributeCode, $data);
     }
 
+    public function createByCheckingData(Attribute $attribute, ?string $channelCode, ?string $localeCode, $data): ValueInterface
+    {
+        if ([] === $data || null === $data) {
+            return $this->createWithoutCheckingData($attribute, $channelCode, $localeCode, []);
+        }
+
+        //QUESTION: SHALL WE TEST IF IT EXISTS ?
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Factory/Read/Value/OptionsValueFactory.php]: to remove

File: src/Akeneo/Pim/Enrichment/Component/Product/Factory/Read/Value/ReferenceDataCollectionValueFactory.php [hunk 3]
```diff
@@ -35,6 +37,26 @@ public function create(Attribute $attribute, ?string $channelCode, ?string $loca
         return ReferenceDataCollectionValue::value($attributeCode, $data);
     }
 
+    public function createByCheckingData(Attribute $attribute, ?string $channelCode, ?string $localeCode, $data): ValueInterface
+    {
+        if ([] === $data || null === $data) {
+            return $this->createWithoutCheckingData($attribute, $channelCode, $localeCode, []);
+        }
+
+        //QUESTION: SHALL WE TEST IF IT EXISTS ?
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Factory/Read/Value/ReferenceDataCollectionValueFactory.php]: to remove

File: src/Akeneo/Pim/Enrichment/Component/Product/Factory/Read/Value/ReferenceDataValueFactory.php [hunk 4]
```diff
@@ -34,6 +35,24 @@ public function create(Attribute $attribute, ?string $channelCode, ?string $loca
         return ReferenceDataValue::value($attributeCode, $data);
     }
 
+    public function createByCheckingData(Attribute $attribute, ?string $channelCode, ?string $localeCode, $data): ValueInterface
+    {
+        if (null === $data) {
+            return $this->createWithoutCheckingData($attribute, $channelCode, $localeCode, $data);
+        }
+
+        //QUESTION: SHALL WE TEST IF IT EXISTS ?
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Factory/Read/Value/ReferenceDataValueFactory.php]: to remove

File: src/Akeneo/Pim/Enrichment/Component/Product/Factory/Read/Value/ValueFactory.php [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface ValueFactory
+{
+    /**
+     * Create the data by validating the data type.
+     * For example, it checks that an expected scalar is a indeed scalar.
+     *
+     * This is done to guarantee that data coming in to the domain are good.
+     * This validation has a non-negligeable cost though.
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Factory/Read/Value/ValueFactory.php]:
```suggestion
     * This validation has a non-negligible cost though.
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Factory/Read/Value/ValueFactory.php [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+     */
+    public function createByCheckingData(Attribute $attribute, ?string $channelCode, ?string $localeCode, $data): ValueInterface;
+
+    /**
+     * Create the data without validating the data type.
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Factory/Read/Value/ValueFactory.php]:
```suggestion
     * Create the value without validating the data type.
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Factory/Read/ValueCollectionFactory.php [hunk 7]
```diff
@@ -141,7 +112,11 @@ private function createValues(array $rawValueCollections, array $attributes): ar
                             $localeCode = null;
                         }
 
-                        $values[] = $this->valueFactory->create($attribute, $channelCode, $localeCode, $data);
+                        try {
+                            //TEMPORARY
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Factory/Read/ValueCollectionFactory.php]: temporary? Could you explain or add a link to a ticket?

File: src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/ImmutableVariantAxesValuesValidator.php [hunk 8]
```diff
@@ -24,12 +24,12 @@ class ImmutableVariantAxesValuesValidator extends ConstraintValidator
     /** @var EntityWithFamilyVariantAttributesProvider */
     private $attributesProvider;
 
-    /** @var WriteValueCollectionFactory */
+    /** @var \Akeneo\Pim\Enrichment\Component\Product\Factory\Read\WriteValueCollectionFactory */
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/ImmutableVariantAxesValuesValidator.php]: useless FQCN

File: src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/ImmutableVariantAxesValuesValidator.php [hunk 9]
```diff
@@ -24,12 +24,12 @@ class ImmutableVariantAxesValuesValidator extends ConstraintValidator
...
     private $valueCollectionFactory;
 
     /**
-     * @param EntityWithFamilyVariantAttributesProvider $attributesProvider
-     * @param WriteValueCollectionFactory           $valueCollectionFactory
+     * @param EntityWithFamilyVariantAttributesProvider                                          $attributesProvider
+     * @param \Akeneo\Pim\Enrichment\Component\Product\Factory\Read\WriteValueCollectionFactory $valueCollectionFactory
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Validator/Constraints/ImmutableVariantAxesValuesValidator.php]: same

File: tests/back/Pim/Enrichment/Specification/Component/Product/Factory/Read/Value/BooleanValueFactorySpec.php [hunk 10]
```diff
@@ -18,63 +20,73 @@ final class BooleanValueFactorySpec extends ObjectBehavior
...
-        $value = $this->create($attribute, null, null, true);
-        $value->isLocalizable()->shouldBe(false);
-        $value->isScopable()->shouldBe(false);
-        $value->getData()->shouldBe(true);
+        $value = $this->createWithoutCheckingData($attribute, null, null, true);
+        $value->shouldBeLike(ScalarValue::value('an_attribute', true));
     }
 
-    public function it_convert_to_boolean_type_the_value()
+    public function it_converts_to_boolean_type_the_value()
```
Comment (mmetayer) [inline tests/back/Pim/Enrichment/Specification/Component/Product/Factory/Read/Value/BooleanValueFactorySpec.php]:
```suggestion
    public function it_converts_the_value_to_boolean_type()
```

File: src/Akeneo/Pim/Enrichment/Component/Product/Factory/Read/WriteValueCollectionFactory.php [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+                        $channelCode = null;
+                    }
+
+                    foreach ($localeRawValue as $localeCode => $data) {
+                        if ('<all_locales>' === $localeCode) {
+                            $localeCode = null;
+                        }
+
+                        try {
+                            //TEMPORARY
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Factory/Read/WriteValueCollectionFactory.php]: same remark

**Pull request merged**